modules/ximgproc/src/l0_smooth.cpp

```diff
@@ -0,0 +1,308 @@
+ *  this list of conditions and the following disclaimer in the documentation
+ *  and / or other materials provided with the distribution.
+ *
+ *  * Neither the names of the copyright holders nor the names of the contributors
+ *  may be used to endorse or promote products derived from this software
+ *  without specific prior written permission.
+ *
+ *  This software is provided by the copyright holders and contributors "as is" and
+ *  any express or implied warranties, including, but not limited to, the implied
+ *  warranties of merchantability and fitness for a particular purpose are disclaimed.
+ *  In no event shall copyright holders or contributors be liable for any direct,
+ *  indirect, incidental, special, exemplary, or consequential damages
+ *  (including, but not limited to, procurement of substitute goods or services;
+ *  loss of use, data, or profits; or business interruption) however caused
+ *  and on any theory of liability, whether in contract, strict liability,
+ *  or tort(including negligence or otherwise) arising in any way out of
+ *  the use of this software, even if advised of the possibility of such damage.
+ */
+
+#include "precomp.hpp"
+#include <vector>
+#include <opencv2/core.hpp>
+#include <opencv2/imgproc.hpp>
+
+using namespace cv;
+using namespace std;
+
+namespace
+{
+	void shift(InputArray src, OutputArray dst, int shift_x, int shift_y) {
+		Mat S = src.getMat();
+		Mat D = dst.getMat();
+
+		if(S.data == D.data){
+			S = S.clone();
+		}
+
+		D.create(S.size(), S.type());
+
+		Mat s0(S, Rect(0, 0, S.cols - shift_x, S.rows - shift_y));
+		Mat s1(S, Rect(S.cols - shift_x, 0, shift_x, S.rows - shift_y));
+		Mat s2(S, Rect(0, S.rows - shift_y, S.cols-shift_x, shift_y));
+		Mat s3(S, Rect(S.cols - shift_x, S.rows- shift_y, shift_x, shift_y));
+
+		Mat d0(D, Rect(shift_x, shift_y, S.cols - shift_x, S.rows - shift_y));
+		Mat d1(D, Rect(0, shift_y, shift_x, S.rows - shift_y));
+		Mat d2(D, Rect(shift_x, 0, S.cols-shift_x, shift_y));
+		Mat d3(D, Rect(0,0,shift_x, shift_y));
+
+		s0.copyTo(d0);
+		s1.copyTo(d1);
+		s2.copyTo(d2);
+		s3.copyTo(d3);
+	}
+
+	// dft after padding imaginary
+	void fft(InputArray src, OutputArray dst) {
+		Mat S = src.getMat();
+		Mat planes[] = {S, Mat::zeros(S.size(), S.type())};
+		merge(planes, 2, dst);
+
+		// compute the result
+		dft(dst, dst);
+	}
+
+	void psf2otf(InputArray src, OutputArray dst, int height, int width){
+		Mat S = src.getMat();
+		Mat D = dst.getMat();
+
+		Mat padded;
+
+		if(S.data == D.data){
+			S = S.clone();
+		}
+
+		// add padding
+		copyMakeBorder(S, padded, 0, height - S.rows, 0, width - S.cols,
+					BORDER_CONSTANT, Scalar::all(0));
+
+		shift(padded, padded, width - S.cols / 2, height - S.rows / 2);
+
+		// convert to frequency domain
+		fft(padded, dst);
+	}
+
+	void dftMultiChannel(InputArray src, vector<Mat> &dst){
+		Mat S = src.getMat();
+
+		split(S, dst);
+
+		for(int i = 0; i < S.channels(); i++){
```

Comment:
* Could be these FFT's done using OpenCV's parallel_for_() & ParallelLoopBody per each channel ?